Implemented "new" keyword for a JSFunction

desertkun opened the pull request:
So you can do this:

```py
from v8py import Context, new

ctx = Context()
ctx.eval("""
    function Test(who) { this.who = who; }
    Test.prototype.test = function() { return "Hello, " + this.who + "!"; }
""")

# Test is a JSFunction here
Test = ctx.glob.Test 

# Javascript equivalent of instance = new Test("world")
instance = new(Test, "world") 

test = instance.test()
print test

> Hello, world!
```

tbodt commented:
I don't like the idea of adding a method to JSFunction or JSObject. What if JavaScript sets `Test.new` to something?

I'd prefer just being able to say `Test()` and have it use the new operator if you're working with a class. This isn't generally possible to find out, though. We could check whether there is a property called `prototype`, but you can use `new` with functions that don't have prototypes. There's a `v8::Object::IsConstructor` method that seems promising, but I'm pretty sure it returns true for every user-defined function (since you can use `new` on those).

Since that won't work, the interface I'd prefer is something like this:

```py
from v8py import new
instance = new(Test, "world")
```
or maybe
```py
instance = new(Test)("world")
```

desertkun commented:
Yeah, good point.

What's next? I can add a new commit into my `bound_function_new` that changes the API into `new(Test, "world")` OR, we can dismiss this pull request and start a new one, can we?

tbodt commented:
You could do either.

desertkun commented:
Okay, changed it to be `context.new(Test, "world")` since context is required. I tried to pull one from `Test`, but context is also required to do get `Local<Value>` of the constructor function in the first place.

Also updated tests.

tbodt commented:
The reason a context is required by `js_from_py` is it has to be able to handle Python objects as well as JavaScript objects, but `new` doesn't need to handle Python objects. In fact, it can throw an exception if the argument isn't a JSFunction. So you can just do `PyArg_ParseTuple(args, "O!", &function, &js_function_type)`. That also takes care of throwing a TypeError if it's the wrong type.

tbodt commented:
Could you implement this as a function on the v8py module?

desertkun commented:
Well, I was going to, but came into a question: what if you need to spawn a new object into context `A` from a constructor function that belongs to context `B`? How the new `new` can handle that? I came into conclusion that explicitly specifying the context would solve the problem.

I can get rid of `js_from_py`, though, and make it to use `js_function_type` like you said, but I have to specify context anyway for the very reason above. Well, I can make function `new(constructor, *args, context=None)` of the v8py module so you can specify one if you want but that's not obvious implementation and this project does lacks documentation. So I am worried would someone ever find this feature?

I also can _additionally_ add a new function into a module that simply uses the context of the constructor function in order to spawn an object. Are you okay with two implementations, one for explicitly specifying target context, and one without such?

What do you say?

tbodt commented:
V8 won't let you call a function or constructor from one context in a different context, the execution always happens in the context where the function was created. You can then put the resulting object in any context you like.

tbodt commented on v8py/utils.cpp:
```diff
@@ -0,0 +1,47 @@
+
+PyObject *construct_new_object(PyObject *self, PyObject *args) {
+
+    int argc = (int)PyTuple_GET_SIZE(args);
```
Why not use PyArg_ParseTuple?

desertkun replied:
I indeed tried to use `PyArg_ParseTuple` with `O!` but it gave me hard time (the js_function* reference it gives was corrupt leading to SIGSEGV) so I ended up with simple `PyTuple_GET_ITEM` / `PyObject_TypeCheck`

tbodt replied:
strange! oh well.

desertkun replied:
Literally the code I've tried:
```
    js_function *function;
    if (PyArg_ParseTuple(args, "O!", &js_function_type, &function) < 0) {
        PyErr_SetString(PyExc_TypeError, "First argument must be a constructor function.");
        return NULL;
    }
```
`function` reference after this execution was way dead.

desertkun commented:
Please take my apology for taking so much of your time regarding this issue.

Finally, moved the method to the function to the v8py module.

tbodt commented:
> Please take my apology for taking so much of your time regarding this issue.

Oh, it's no problem.
